src/include/stir/KOSMAPOSL/KOSMAPOSLReconstruction.h

```diff
@@ -0,0 +1,259 @@
+  sigma dp:=1                                ;is the parameter $\f \sigma_{dp} \f$;
+  number of neighbours:= 3                   ;is the cubic root of the number of voxels in the neighbourhood;
+  anatomical image filename:=filename        ;is the filename of the anatomical image;
+  number of non-zero feature elements:=1     ;is the number of non zero elements in the feature vector;
+  only_2D:=0                                 ;=1 if you want to reconstruct 2D images;
+
+
+  End KOSMAPOSL Parameters :=
+  \endverbatim
+*/
+
+template <typename TargetT>
+class KOSMAPOSLReconstruction:
+        public
+            RegisteredParsingObject<
+        KOSMAPOSLReconstruction <TargetT > ,
+        Reconstruction < TargetT >,
+        OSMAPOSLReconstruction < TargetT >
+         >
+
+//public IterativeReconstruction<TargetT >
+{
+ private:
+  typedef RegisteredParsingObject<
+    KOSMAPOSLReconstruction <TargetT > ,
+    Reconstruction < TargetT >,
+    OSMAPOSLReconstruction < TargetT >
+     >
+
+
+    base_type;
+public:
+
+    //! Name which will be used when parsing a KOSMAPOSLReconstruction object
+    static const char * const registered_name;
+
+  //! Default constructor (calling set_defaults())
+  KOSMAPOSLReconstruction();
+  /*!
+  \brief Constructor, initialises everything from parameter file, or (when
+  parameter_filename == "") by calling ask_parameters().
+  */
+  explicit 
+    KOSMAPOSLReconstruction(const std::string& parameter_filename);
+
+  //! accessor for the external parameters
+  KOSMAPOSLReconstruction& get_parameters(){return *this; }
+
+  //! accessor for the external parameters
+  const KOSMAPOSLReconstruction& get_parameters() const
+    {return *this;}
+
+  //kernel
+  const std::string get_anatomical_filename() const;
+  const int get_num_neighbours() const;
+  const int get_num_non_zero_feat() const;
+  const double get_sigma_m() const;
+  double get_sigma_p();
+  double get_sigma_dp();
+  double get_sigma_dm();
+  const bool get_only_2D() const;
+  bool get_hybrid();
+
+   shared_ptr<TargetT>& get_kpnorm_sptr();
+   shared_ptr<TargetT>& get_kmnorm_sptr();
+   shared_ptr<TargetT>& get_anatomical_sptr();
+
+    /*! \name Functions to set parameters
+    This can be used as alternative to the parsing mechanism.
+   \warning Be careful with setting shared pointers. If you modify the objects in 
+   one place, all objects that use the shared pointer will be affected.
+  */
+
+  void set_kpnorm_sptr(shared_ptr<TargetT>&);
+  void set_kmnorm_sptr(shared_ptr<TargetT>&);
+  void set_anatomical_sptr(shared_ptr<TargetT>&);
+
+
+  //! boolean value to determine if the update images have to be written to disk
+  void set_write_update_image(const int);
+
+
+  //! prompts the user to enter parameter values manually
+  virtual void ask_parameters();
+
+ protected:
+  //! Filename with input projection data
+  std::string input_filename,kernelised_output_filename_prefix;
+  std::string current_kimage_filename;
```

Comment:
I had thought that too when we discussed but then I tried and the results are different. This is because we thought that KOSMAPOSL would just use the lambda image, but in this way the kernel will depend on lambda instead of alpha, it is like writing lambda=f(lambda) instead of  lambda=f(alpha). It is necessary to apply the extra step if we want the PET image, that is why we can avoid printing the alpha from the parfile and just print the PET image using the "current_kimage_filename".

Reply:
Hmmm, I think I see the issue. The proposal was essentially to "mask out" alpha and pass lambda back and forth. But we need alpha to calculate the kernel, and also to apply the update to. Perhaps we can hide alpha inside KOSMAPOSL as some state. But then the `update_estimate` works a little differently to how OSMAPOSL uses it.